### PR TITLE
Bitstamp: Add new trading fees API endpoint; refine fee calcuations and test coverage

### PR DESCRIPTION
**Pull request by Beadko**
# PR Description
A fix for issue #1271 
The trading fees have not been updated, always being 0.
Upon review, it showed a change in the trading fee fetching between v1 and v2.

Fixes # (issue)
Get trading fees directly from the API
## Type of change

Please delete options that are not relevant and add an `x` in `[]` as item is complete.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How has this been tested

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration and
also consider improving test coverage whilst working on a certain feature or package.

- [x] go test ./... -race
- [x] golangci-lint run
- [x] TestGetFee

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally and on Github Actions/AppVeyor with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: exchanges/bitstamp/bitstamp.go [hunk 1]
```diff
@@ -86,6 +83,23 @@ func (b *Bitstamp) GetFee(ctx context.Context, feeBuilder *exchange.FeeBuilder)
 	return fee, nil
 }
 
+// GetTradingFee returns a trading fee based on a currency
+func (b *Bitstamp) getTradingFee(ctx context.Context, feeBuilder *exchange.FeeBuilder) (float64, error) {
+	var resp TradingFees
+	err := b.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, bitstampAPITradingFees+"/"+strings.ToLower(feeBuilder.Pair.String()), true, nil, &resp)
+	if err != nil {
+		return 0, err
+	}
```
Comment (shazbert) [inline exchanges/bitstamp/bitstamp.go:92]: Could you please make a separate function for this that returns a slice of TradingFees struct? And allow for an empty currency pair so it returns the entire list of fees for all trading pairs. 

```go
func (b *Bitstamp) GetAccountTradingFees(ctx context.Context, pair currency.Pair) ([]TradingFees, error) {}
```

That way a client can just use this function without having to use the feeBuilder functionality if they chose not too.

**Review by shazbert — CHANGES_REQUESTED** [review]
Looking good 💃. Just one request. Thanks Bea.

File: exchanges/bitstamp/bitstamp.go (hunk 1 above)
Comment (Beadko) [reply exchanges/bitstamp/bitstamp.go:92]: Let's make it a separate PR as ideally, it should be an interface function to cover all the exchanges to avoid the divergence between the exchanges. What do you think?

Comment (shazbert) [reply exchanges/bitstamp/bitstamp.go:92]: Just a library function in `bitstamp.go` `GetAccountTradingFees` should suffice, agreed we can make an interface function for it later in a PR.

Comment (Beadko) [reply exchanges/bitstamp/bitstamp.go]: function added

File: exchanges/bitstamp/bitstamp.go [hunk 2]
```diff
@@ -86,6 +83,40 @@ func (b *Bitstamp) GetFee(ctx context.Context, feeBuilder *exchange.FeeBuilder)
...
+}
+
+// GetAccountTradingFees returns a slice of TradingFee
+func (b *Bitstamp) GetAccountTradingFees(ctx context.Context) ([]TradingFees, error) {
+	path := bitstampAPITradingFees + "/"
+
+	var resp []TradingFees
+	err := b.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, path, true, nil, &resp)
+	return resp, err
+}
```
Comment (shazbert) [inline exchanges/bitstamp/bitstamp.go:118]: Can be merged into one function, also the empty pair case will need to be tested, there was missing test coverage for `GetAccountTradingFees`:
```go
// GetAccountTradingFees returns account tradings fee for a pair or all fees if pair is empty.
func (b *Bitstamp) GetAccountTradingFees(ctx context.Context, pair currency.Pair) ([]TradingFees, error) {
	path := bitstampAPITradingFees + "/" + strings.ToLower(pair.String())
	if pair.IsEmpty() {
		var resp []TradingFees
		err := b.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, path, true, nil, &resp)
		return resp, err
	}
	var resp TradingFees
	err := b.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, path, true, nil, &resp)
	return []TradingFees{resp}, err
}

```

**Review by shazbert — CHANGES_REQUESTED** [review]
One last thing from me then looks good to go. Thanks 🕺

File: exchanges/bitstamp/bitstamp.go (hunk 2 above)
Comment (gbjk) [reply exchanges/bitstamp/bitstamp.go:118]: @shazbert  What I don't like about this is that `GetAccountTradingFee` is our only current use-case.

Adding `GetAccountTradingFees` is a nice-to-have, but it doesn't fill a current interface gap or use-case we have now.

So when we're making our func API harder to use for our only use-case, with the only benefit being that there's one func instead of 2, it worries me.

Now the consumer of `GetAccountTradingFees` needs to check it gets back only one, and also pull `*.Fees` out.
Arguably I'd move that check, and also the check of only 1 .Fees into `GetAccountTradingFee` to further simplify the API for consumers.

Finally, I think that having an empty pair is slightly distasteful. I'd probably use a variadic.

TL;DR: I don't think we should be afraid to have 2 funcs if it improves usability and API.
Just wanted to voice my opinion. No skin in the game.

Comment (shazbert) [reply exchanges/bitstamp/bitstamp.go:118]: I tend to always default to consolidate calls with similar types and outbound request paths especially in this file, align one function to one endpoint for direct API mapping. 

I do see your point though to make it easier for users but that is why we have a common approach through the wrapper which will handle the heaving lifting of consuming this function call.

Look either way 2 functions doesn't hurt either 🤷. @gloriousCode @thrasher- Thoughts?

File: exchanges/bitstamp/bitstamp.go [hunk 3]
```diff
@@ -86,6 +83,40 @@ func (b *Bitstamp) GetFee(ctx context.Context, feeBuilder *exchange.FeeBuilder)
...
+	}
+	fee := fees[0].Taker
+	if feeBuilder.IsMaker {
+		fee = fees[0].Maker
+	}
+	return fee / 100 * feeBuilder.PurchasePrice * feeBuilder.Amount, nil
+}
+
+// GetAccountTradingFee returns a TradingFee for a pair
+func (b *Bitstamp) GetAccountTradingFee(ctx context.Context, pair currency.Pair) ([]MakerTakerFees, error) {
```
Comment (gloriousCode) [inline exchanges/bitstamp/bitstamp.go:103]: This endpoint does not work. Please add a test
```
    bitstamp_test.go:874: json: cannot unmarshal object into Go struct field TradingFees.Fees of type []bitstamp.MakerTakerFees
    bitstamp_test.go:877: []
```

File: exchanges/bitstamp/bitstamp.go [hunk 4]
```diff
@@ -86,6 +83,40 @@ func (b *Bitstamp) GetFee(ctx context.Context, feeBuilder *exchange.FeeBuilder)
...
+	path := bitstampAPITradingFees + "/" + strings.ToLower(pair.String())
+
+	var resp TradingFees
+	err := b.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, path, true, nil, &resp)
+	return resp.Fees, err
+}
+
+// GetAccountTradingFees returns a slice of TradingFee
+func (b *Bitstamp) GetAccountTradingFees(ctx context.Context) ([]TradingFees, error) {
```
Comment (gloriousCode) [inline exchanges/bitstamp/bitstamp.go:112]: This endpoint returns errors, please add a test
```
Bitstamp raw response: [{"currency_pair": "1incheur", "market": "1incheur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "1inchusd", "market": "1inchusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "aavebtc", "market": "aavebtc", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "aaveeur", "market": "aaveeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "aaveusd", "market": "aaveusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "adabtc", "market": "adabtc", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "adaeur", "market": "adaeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "adausd", "market": "adausd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "algobtc", "market": "algobtc", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "algoeur", "market": "algoeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "algousd", "market": "algousd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "alphaeur", "market": "alphaeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "alphausd", "market": "alphausd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "ampeur", "market": "ampeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "ampusd", "market": "ampusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "anteur", "market": "anteur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "antusd", "market": "antusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "apeeur", "market": "apeeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "apeusd", "market": "apeusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "audiobtc", "market": "audiobtc", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "audioeur", "market": "audioeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "audiousd", "market": "audiousd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "avaxeur", "market": "avaxeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "avaxusd", "market": "avaxusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "axseur", "market": "axseur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "axsusd", "market": "axsusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "bandeur", "market": "bandeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "bandusd", "market": "bandusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "bateur", "market": "bateur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "batusd", "market": "batusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "bchbtc", "market": "bchbtc", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "bcheur", "market": "bcheur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "bchusd", "market": "bchusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "btceur", "market": "btceur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "btcgbp", "market": "btcgbp", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "btcpax", "market": "btcpax", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "btcusd", "market": "btcusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "btcusdc", "market": "btcusdc", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "btcusdt", "market": "btcusdt", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "chzeur", "market": "chzeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "chzusd", "market": "chzusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "compeur", "market": "compeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "compusd", "market": "compusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "crveur", "market": "crveur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "crvusd", "market": "crvusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "ctsieur", "market": "ctsieur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "ctsiusd", "market": "ctsiusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "cvxeur", "market": "cvxeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "cvxusd", "market": "cvxusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "daiusd", "market": "daiusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "dgldeur", "market": "dgldeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "dgldusd", "market": "dgldusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "dogeeur", "market": "dogeeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "dogeusd", "market": "dogeusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "doteur", "market": "doteur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "dotusd", "market": "dotusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "dydxeur", "market": "dydxeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "dydxusd", "market": "dydxusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "enjeur", "market": "enjeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "enjusd", "market": "enjusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "enseur", "market": "enseur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "ensusd", "market": "ensusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "eth2eth", "market": "eth2eth", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "ethbtc", "market": "ethbtc", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "etheur", "market": "etheur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "ethgbp", "market": "ethgbp", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "ethpax", "market": "ethpax", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "ethusd", "market": "ethusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "ethusdc", "market": "ethusdc", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "ethusdt", "market": "ethusdt", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "euroceur", "market": "euroceur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "eurocusdc", "market": "eurocusdc", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "eurteur", "market": "eurteur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "eurtusd", "market": "eurtusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "eurusd", "market": "eurusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "feteur", "market": "feteur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "fetusd", "market": "fetusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "flreur", "market": "flreur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "flrusd", "market": "flrusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "ftmeur", "market": "ftmeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "ftmusd", "market": "ftmusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "galaeur", "market": "galaeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "galausd", "market": "galausd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "gbpusd", "market": "gbpusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "godseur", "market": "godseur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "godsusd", "market": "godsusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "grteur", "market": "grteur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "grtusd", "market": "grtusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "gusdusd", "market": "gusdusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "hbareur", "market": "hbareur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "hbarusd", "market": "hbarusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "imxeur", "market": "imxeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "imxusd", "market": "imxusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "injeur", "market": "injeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "injusd", "market": "injusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "knceur", "market": "knceur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "kncusd", "market": "kncusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "ldoeur", "market": "ldoeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "ldousd", "market": "ldousd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "linkbtc", "market": "linkbtc", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "linkeur", "market": "linkeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "linkgbp", "market": "linkgbp", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "linkusd", "market": "linkusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "lrceur", "market": "lrceur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "lrcusd", "market": "lrcusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "ltcbtc", "market": "ltcbtc", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "ltceur", "market": "ltceur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "ltcgbp", "market": "ltcgbp", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "ltcusd", "market": "ltcusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "manaeur", "market": "manaeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "manausd", "market": "manausd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "maticeur", "market": "maticeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "maticusd", "market": "maticusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "mkreur", "market": "mkreur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "mkrusd", "market": "mkrusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "mpleur", "market": "mpleur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "mplusd", "market": "mplusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "neareur", "market": "neareur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "nearusd", "market": "nearusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "nexoeur", "market": "nexoeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "nexousd", "market": "nexousd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "paxusd", "market": "paxusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "perpeur", "market": "perpeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "perpusd", "market": "perpusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "radeur", "market": "radeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "radusd", "market": "radusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "rlyeur", "market": "rlyeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "rlyusd", "market": "rlyusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "rndreur", "market": "rndreur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "rndrusd", "market": "rndrusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "sandeur", "market": "sandeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "sandusd", "market": "sandusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "sgbeur", "market": "sgbeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "sgbusd", "market": "sgbusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "shibeur", "market": "shibeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "shibusd", "market": "shibusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "skleur", "market": "skleur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "sklusd", "market": "sklusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "slpeur", "market": "slpeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "slpusd", "market": "slpusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "snxeur", "market": "snxeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "snxusd", "market": "snxusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "soleur", "market": "soleur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "solusd", "market": "solusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "storjeur", "market": "storjeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "storjusd", "market": "storjusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "suieur", "market": "suieur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "suiusd", "market": "suiusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "sushieur", "market": "sushieur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "sushiusd", "market": "sushiusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "sxpeur", "market": "sxpeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "sxpusd", "market": "sxpusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "umaeur", "market": "umaeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "umausd", "market": "umausd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "unibtc", "market": "unibtc", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "unieur", "market": "unieur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "uniusd", "market": "uniusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "usdceur", "market": "usdceur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "usdcusd", "market": "usdcusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "usdcusdt", "market": "usdcusdt", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "usdteur", "market": "usdteur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "usdtusd", "market": "usdtusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "vegaeur", "market": "vegaeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "vegausd", "market": "vegausd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "wbtcbtc", "market": "wbtcbtc", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "xlmbtc", "market": "xlmbtc", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "xlmeur", "market": "xlmeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "xlmgbp", "market": "xlmgbp", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "xlmusd", "market": "xlmusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "xrpbtc", "market": "xrpbtc", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "xrpeur", "market": "xrpeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "xrpgbp", "market": "xrpgbp", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "xrpusd", "market": "xrpusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "xrpusdt", "market": "xrpusdt", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "yfieur", "market": "yfieur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "yfiusd", "market": "yfiusd", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "zrxeur", "market": "zrxeur", "fees": {"maker": "0.000", "taker": "0.000"}}, {"currency_pair": "zrxusd", "market": "zrxusd", "fees": {"maker": "0.000", "taker": "0.000"}}]
    bitstamp_test.go:886: json: cannot unmarshal object into Go struct field TradingFees.Fees of type []bitstamp.MakerTakerFees
```

File: exchanges/bitstamp/bitstamp.go (hunk 2 above)
Comment (gloriousCode) [reply exchanges/bitstamp/bitstamp.go:118]: The goal of {{exchange}}.go is to implement the API endpoints as close to documentation as possible. A user can use an exchange's package and know that they can call the functions that are described in the API docs. Helper functions are allowed too. Given that, an individual's desired use-case is not really a consideration for a public repository. Checking the length of a response is hardly a problem, nor is having empty params, otherwise you'll be rewriting all {{exchange}}.go implementations.

With that goal in mind, [trading fees](https://www.bitstamp.net/api/#tag/Fees/operation/GetAllTradingFees) and [trading fee for currency](https://www.bitstamp.net/api/#tag/Fees/operation/GetTradingFeesForCurrency) are two separate entries and so will be fine to have as two

File: exchanges/bitstamp/bitstamp.go [hunk 5]
```diff
@@ -86,6 +83,40 @@ func (b *Bitstamp) GetFee(ctx context.Context, feeBuilder *exchange.FeeBuilder)
...
+	return fee / 100 * feeBuilder.PurchasePrice * feeBuilder.Amount, nil
+}
+
+// GetAccountTradingFee returns a TradingFee for a pair
+func (b *Bitstamp) GetAccountTradingFee(ctx context.Context, pair currency.Pair) ([]MakerTakerFees, error) {
+	path := bitstampAPITradingFees + "/" + strings.ToLower(pair.String())
+
+	var resp TradingFees
+	err := b.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, path, true, nil, &resp)
+	return resp.Fees, err
```
Comment (gloriousCode) [inline exchanges/bitstamp/bitstamp.go:108]: Please return `resp`, not `resp.Fees` to not filter out response API data

**Review by gloriousCode — CHANGES_REQUESTED** [review]
Thank you for updating the trading fee endpoint!

File: exchanges/bitstamp/bitstamp.go (hunk 2 above)
Comment (Beadko) [reply exchanges/bitstamp/bitstamp.go]: Noted, thank you @gloriousCode

File: exchanges/bitstamp/bitstamp.go (hunk 5 above)
Comment (Beadko) [reply exchanges/bitstamp/bitstamp.go]: Completed in b0f9d3

File: exchanges/bitstamp/bitstamp.go (hunk 3 above)
Comment (Beadko) [reply exchanges/bitstamp/bitstamp.go]: Noted, working on it

File: exchanges/bitstamp/bitstamp.go (hunk 4 above)
Comment (Beadko) [reply exchanges/bitstamp/bitstamp.go:113]: Noted, working on it

**Review by shazbert — CHANGES_REQUESTED** [review]
Just waiting on this [nit to be actioned](https://github.com/thrasher-corp/gocryptotrader/pull/1289/files#r1329418949) RE: The additional test.

File: exchanges/bitstamp/bitstamp.go [hunk 6]
```diff
@@ -86,6 +83,36 @@ func (b *Bitstamp) GetFee(ctx context.Context, feeBuilder *exchange.FeeBuilder)
...
+	fees := tradingFees.Fees
+	fee := fees.Taker
+	if feeBuilder.IsMaker {
+		fee = fees.Maker
+	}
+	return fee / 100 * feeBuilder.PurchasePrice * feeBuilder.Amount, nil
+}
+
+// GetAccountTradingFee returns a TradingFee for a pair
+func (b *Bitstamp) GetAccountTradingFee(ctx context.Context, pair currency.Pair) (TradingFees, error) {
```
Comment (shazbert) [inline exchanges/bitstamp/bitstamp.go:101]: If I pass in an empty currency pair it returns a slice, so maybe make a check if the currency pair is empty and return an error. 
```
json: cannot unmarshal array into Go value of type bitstamp.TradingFees
```

**Review by shazbert — CHANGES_REQUESTED** [review]
Please run go mod tidy and then one nit from me. Thanks for the updated tests.

File: exchanges/bitstamp/bitstamp_test.go [hunk 7]
```diff
@@ -140,28 +154,45 @@ func TestGetFee(t *testing.T) {
...
+	if !mockTests {
+		sharedtestvalues.SkipTestIfCredentialsUnset(t, b)
 	}
 
-	dummy1, dummy2 := currency.NewCode(""), currency.NewCode("")
-	if resp := b.CalculateTradingFee(dummy1, dummy2, 0, 0, newBalance); resp != 0 {
-		t.Error("GetFee() error")
+	fees, err := b.GetAccountTradingFees(context.Background())
+	if assert.NoError(t, err, "GetAccountTradingFee should not error") {
+
```
Comment (shazbert) [inline exchanges/bitstamp/bitstamp_test.go:187]: The linter is complaining about this empty line, please RM and the test should pass:

https://ci.appveyor.com/project/thrasher-/gocryptotrader-vxsws/builds/48230331#L299

**Review by shazbert — CHANGES_REQUESTED** [review]
Thanks RE: linter

File: exchanges/bitstamp/bitstamp_test.go (hunk 7 above)
Comment (Beadko) [reply exchanges/bitstamp/bitstamp_test.go]: done, thank you :)

**Review by shazbert — APPROVED** [review]
LGTM Thanks BEA 👍

**Review by gloriousCode — CHANGES_REQUESTED** [review]
I got a request to `go mod tidy`, could you do that again? I do see that the tests ran last time though..
If nothing changes, I'll approve and blame my machine 🎉

**Comment by gloriousCode** [comment]
Sorry @Beadko , there's now a merge conflict, once merged, I'll approve

**Comment by Beadko** [comment]
> Sorry @Beadko , there's now a merge conflict, once merged, I'll approve

The merge conflict has been solved

**Review by gloriousCode — APPROVED** [review]
tACK!

**Review by shazbert — APPROVED** [review]
changes ACK. Thanks BEA.

**Review by thrasher- — APPROVED** [review]
Thanks @Beadko 🚀
